File: bridges/go-python/tests/test-child.py
"""
test-child.py - Stitch test sidecar for the Go → Python bridge.

Methods exposed
---------------
echo          {"msg": str}                → {"echo": str}
add           {"a": number, "b": number}  → {"sum": number}
raise_error   {}                          → RPC error response
echo_b64      {"data": str}              → {"data": str}  (round-trip base64)
slow          {"ms": int}                 → {"slept_ms": int}
"""
import sys
_rpc_out = sys.stdout
sys.stdout = sys.stderr

import base64
import json
import logging
import os
import threading
import time
import traceback

logging.disable(logging.CRITICAL)

# ---------------------------------------------------------------------------
# Watchdog
# ---------------------------------------------------------------------------

def _stdin_watchdog() -> None:
    sys.stdin.read()
    os._exit(0)

threading.Thread(target=_stdin_watchdog, daemon=True).start()

# ---------------------------------------------------------------------------
# Handlers
# ---------------------------------------------------------------------------

def _handle_echo(params: dict):
    return {"echo": params.get("msg", "")}


def _handle_add(params: dict):
    a = params["a"]
    b = params["b"]
    return {"sum": a + b}


def _handle_raise_error(params: dict):
    raise ValueError("intentional test error from Python sidecar")


def _handle_echo_b64(params: dict):
    # Round-trip: decode then re-encode to prove we touched it.
    raw = base64.b64decode(params["data"])
    return {"data": base64.b64encode(raw).decode()}


def _handle_slow(params: dict):
    ms = int(params.get("ms", 100))
    time.sleep(ms / 1000.0)
    return {"slept_ms": ms}


HANDLERS = {
    "echo": _handle_echo,
    "add": _handle_add,
    "raise_error": _handle_raise_error,
    "echo_b64": _handle_echo_b64,
    "slow": _handle_slow,
}

# ---------------------------------------------------------------------------
# RPC machinery
# ---------------------------------------------------------------------------

def _send(obj: dict) -> None:
    _rpc_out.write(json.dumps(obj) + "\n")
    _rpc_out.flush()


def _dispatch(line: str) -> None:
    try:
        req = json.loads(line)
    except json.JSONDecodeError as exc:
        print(f"[test-child] JSON parse error: {exc}", file=sys.stderr)
        return

    req_id = req.get("id", "")
    method = req.get("method", "")
    params = req.get("params") or {}

    handler = HANDLERS.get(method)
    if handler is None:
        _send({
            "id": req_id,
            "error": {"message": f"Unknown method: {method!r}", "traceback": ""},
        })
        return

    try:
        result = handler(params)
        _send({"id": req_id, "result": result})
    except Exception as exc:
        _send({
            "id": req_id,
            "error": {"message": str(exc), "traceback": traceback.format_exc()},
        })


def main() -> None:
    _send({"ready": True})
    for raw_line in sys.stdin:
        line = raw_line.strip()
        if not line:
            continue
        _dispatch(line)


if __name__ == "__main__":
    main()

File: bridges/go-nodejs/tests/test-child.js
// test-child.js - real Node.js sidecar used by the Go test suite.
//
// Methods:
//   echo        - returns params unchanged
//   add         - {a, b} → {sum}
//   raise_error - always returns a JSON-RPC error
//   echo_b64    - decodes a base-64 string and re-encodes it (Buffer round-trip)
//   slow        - waits `ms` milliseconds then returns {tag, ms}
'use strict';

const readline = require('readline');

const handlers = {
  echo: async (params) => params,

  add: async ({ a, b }) => ({ sum: a + b }),

  raise_error: async ({ msg }) => {
    throw new Error(msg || 'intentional error');
  },

  echo_b64: async ({ data }) => {
    // Decode base-64 → Buffer → re-encode as base-64.
    const buf = Buffer.from(data, 'base64');
    return { decoded: buf.toString('utf8'), reencoded: buf.toString('base64') };
  },

  slow: async ({ ms = 100, tag = 0 }) => {
    await new Promise((resolve) => setTimeout(resolve, ms));
    return { tag, ms };
  },
};

async function dispatch(req) {
  const handler = handlers[req.method];
  if (!handler) {
    return {
      id: req.id,
      error: { code: -32601, message: `Method not found: ${req.method}` },
    };
  }
  try {
    const result = await handler(req.params || {});
    return { id: req.id, result };
  } catch (err) {
    return {
      id: req.id,
      error: { code: -32000, message: err.message || String(err) },
    };
  }
}

const rl = readline.createInterface({ input: process.stdin });

process.stdout.write(JSON.stringify({ ready: true }) + '\n');

rl.on('line', async (line) => {
  const trimmed = line.trim();
  if (!trimmed) return;
  let req;
  try {
    req = JSON.parse(trimmed);
  } catch {
    return;
  }
  const resp = await dispatch(req);
  process.stdout.write(JSON.stringify(resp) + '\n');
});

rl.on('close', () => process.exit(0));

process.on('SIGTERM', () => process.exit(0));
process.on('SIGINT',  () => process.exit(0));
